File: app/src/main/kotlin/com/victor/githubclient/loader/GitHubLoaderManager.java
package com.victor.githubclient.loader;

import android.support.v4.app.LoaderManager;
import android.support.v4.app.LoaderManager.LoaderCallbacks;
import android.os.Bundle;

public class GitHubLoaderManager {

    public static <D> void init(final LoaderManager manager, final int loaderId,
                                   final GitHubLoader<D> loader, final Callback<D> callback) {
        manager.initLoader(loaderId, Bundle.EMPTY, new LoaderCallbacks<Response<D>>() {
            @Override
            public android.support.v4.content.Loader<Response<D>> onCreateLoader(int id, Bundle args) {
                return loader;
            }

            @Override
            public void onLoadFinished(android.support.v4.content.Loader<Response<D>> loader, Response<D> data) {
                if (data.hasError()) {
                    callback.onFailure(data.getException());
                } else {
                    callback.onSuccess(data.getResult());
                }

                manager.destroyLoader(loaderId);
            }

            @Override
            public void onLoaderReset(android.support.v4.content.Loader<Response<D>> loader) {
                //Nothing to do here
            }
        });
    }
}

File: app/src/main/kotlin/com/victor/githubclient/loader/GitHubLoaderManager.kt
package com.victor.githubclient.loader

import android.os.Bundle
import android.support.v4.app.LoaderManager
import android.support.v4.app.LoaderManager.LoaderCallbacks

object GitHubLoaderManager {
    fun <T> init(manager: LoaderManager, loaderId: Int,
                 loader: GitHubLoader<T>, callback: Callback<T>) {
        manager.initLoader(loaderId, Bundle.EMPTY, object : LoaderCallbacks<Response<T>> {
            override fun onCreateLoader(id: Int, args: Bundle): android.support.v4.content.Loader<Response<T>> {
                return loader
            }

            override fun onLoadFinished(loader: android.support.v4.content.Loader<Response<T>>, data: Response<T>) {
                data.result?.let { callback.onSuccess(it) }
                data.exception?.let { callback.onFailure(it) }

                manager.destroyLoader(loaderId)
            }

            override fun onLoaderReset(loader: android.support.v4.content.Loader<Response<T>>) {
                //Nothing to do here
            }
        })
    }
}
